actionpack/lib/action_controller/metal/renderers.rb

```diff
@@ -18,16 +18,33 @@ def initialize(format)
     end
   end
 
+  class MissingSerializingFormat < KeyError
+    def initialize(format)
+      super "No serializing format defined for: '#{format}'."
+    end
+  end
+
   module Renderers
     extend ActiveSupport::Concern
 
     # A Set containing renderer names that correspond to available renderer procs.
     # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
     RENDERERS = Set.new
+    FORMATTERS = Hash.new do |h, format|
+        fail ActionController::MissingSerializingFormat.new(format)
+    end
 
     included do
       class_attribute :_renderers
       self._renderers = Set.new.freeze
+      class_attribute :_formatters
+      self._formatters = FORMATTERS.dup.freeze
+    end
+
+    class_methods do
+      def serializing(**formatters)
+        self._formatters = _formatters.merge(formatters)
```

Comment:
@dhh @rafaelfranca I implemented `serializing` to use a `_formatters` class_attribute which maps a format to its formatter.  (avoiding the word 'serializer'). (per https://github.com/rails/rails/pull/21496#discussion_r53316942 )

I've chosen to implement it in the least invasive way, which is to add a 'formatter' to the json/js/xml rendrerers, but not require it for the renderer to work, since the response body is set from the return value of the renderer in any case.

Other thoughts I've had are to make the 'RENDERERS' behave like the 'FORMATTERS', since formatters are 1->1. (but renderers are 1->0,1), but that makes some changes such as `use_renderer` more difficult.

I also considered adding the 'formatter' as a second argument to the renderer, but there wasn't a big advantage to that.

I'm interested in your thoughts this implementation, interface, and testing.

At this time, it is possible to add a formatter with no corresponding renderer. It just won't be used.  There's also not a clear public interface for removing a 'formatter'.  I'm not sure there should be one.